library/cwm/src/lib/cwm/progress_bar.rb

```diff
@@ -0,0 +1,207 @@
+# Copyright (c) [2020] SUSE LLC
+#
+# All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of version 2 of the GNU General Public License as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "cwm/custom_widget"
+
+module CWM
+  # Widget for a progress bar
+  #
+  # @example
+  #
+  #   pg = ProgressBar.new(steps: ["action 1", "action 2"])
+  #   pg.forward #=> shows label "action 1"
+  #   pg.forward #=> shows label "action 2"
+  #
+  #   pg = ProgressBar.new(steps: ["action 1", "action 2"], label: "Progress")
+  #   pg.forward #=> shows label "action 1"
+  #   pg.forward #=> shows label "action 2"
+  #
+  #   pg = ProgressBar.new(steps: ["action 1", "action 2"], label: "Progress", show_step: false)
+  #   pg.forward #=> shows label "Progress"
+  #   pg.forward #=> shows label "Progress"
+  #
+  #   pg = ProgressBar.new(total_steps: 3, label: "Progress")
+  #   pg.forward                    #=> shows label "Progress"
+  #   pg.forward(label: "Action 2") #=> shows label "Action 2"
+  #   pg.forward                    #=> shows label "Progress"
+  #
+  #   pg = ProgressBar.new(steps: ["action 1", "action 2", "action 3"], label: "Progress")
+  #   pg.forward                    #=> shows label "action 1"
+  #   pg.forward(label: "Action X") #=> shows label "Action X"
+  #   pg.forward                    #=> shows label "action 3"
+  class ProgressBar < CustomWidget
+    # Constructor
+    #
+    # @param id [String] widget id; auto-generated if nothing is indicated.
+    # @param label [String] label for the progress bar. This is ignored in case that steps are known (see
+    #   steps parameter) and the steps are configured to be shown (see show_step parameter).
+    # @param total_steps [Integer] total number of steps that the progress bar can perform. This is
+    #   ignored in case that steps are know (see steps parameter).
+    # @param steps [Array<String>] list of steps. They will be used as label of the progress bar in case
+    #   steps are configured to be shown (see show_step parameter).
+    # @param show_step [Boolean] whether the current step should be used as label for the progreess bar.
+    #   This is ignored if the steps are unknown (see steps parameter).
+    def initialize(id: nil, label: nil, total_steps: 0, steps: [], show_step: true)
```

Comment:
well, API style here is kind of different to CWM style. Here it is more like instantiate object and calling actions there. CWM way is more like specialized object ( usually child ) and then calling actions on it.

Difference is constructor. Usual CWM constructor is almost empty and just depends on it defined methods.

What I also do not like at this code is unnecessary complexity that mixture two styles. One is defined steps and one is undefined steps. So what it can look like?
There can be two classes one is ProgressBar which subclass has to define steps method and optional show_step method. And its subclass which can be called DynamicProgressBar, that require label method and steps_count and define steps as label * steps_count.
This will also solve all issues with id that need not to be autogenerated. So in the end it can be empty constructor and two classes. Where you pick behavior by class and not by combination of parameters in constructor.